### ISSUE TITLE
Player does not work when captions are unavailables on Youtube

### ISSUE THREAD

**Issue by DuaelFr** [ableplayer/ableplayer#554]
I've been struggling for quite some time with what seemed to be a random bug.

As you can see on [this page](https://www.college-de-france.fr/agenda/colloque/tabou-et-transgressions/la-question-de-interdit-du-porc-en-egypte-ancienne) the video player is loaded but no controls are available whereas on [this other page](https://www.college-de-france.fr/agenda/colloque/tabou-et-transgressions/le-sabbat-et-le-sang-propos-de-deux-tabous-majeurs-du-judaisme-epoque-perse) it works perfectly.
Without errors in the browser console I had to dig a lot to understand that in [the first case](https://www.youtube.com/watch?v=r13a8d5YGFY), captions are not available at all (not even auto-generated ones so the button is disabled on the Youtube player), while in [the second case](https://www.youtube.com/watch?v=7zOZyUDNF8g) auto-generated captions are available (on Youtube only, AblePlayer doesn't seem to support them for now).

I hope this gives you enough context to find a fix!
Feel free to ask me if you need more info.

**Comment by xerc** [comment]
remove `data-hide-controls="false"` and add `playsinline`
btw/ there is no need for `="true"` @ `data-youtube-nocookie`

**Comment by DuaelFr** [comment]
Thank you @xerc for your suggestions but it does not fix the issue.

**Comment by xerc** [comment]
there is no change @ [first-link](https://www.college-de-france.fr/agenda/colloque/tabou-et-transgressions/la-question-de-interdit-du-porc-en-egypte-ancienne)

**Comment by xerc** [comment]
caption-functions do not behave "normal" with YT ; https://github.com/ableplayer/ableplayer/issues/562

**Comment by drumwhacker** [comment]
Having this issue as well.  Did you decide not to use the player?

**Comment by DuaelFr** [comment]
Sadly we had to remove AblePlayer for Youtube videos because of this issue. We kept it for self hosted ones and for audio content though.

**Comment by drumwhacker** [comment]
Thank you for your answer, Edouard!  We are using this player too much now
to just dismiss it I'll need to find some way to address this.  If it's
anything workable for someone else, I will certainly share it.

-Claire

**Comment by terrill** [comment]
Hi All.  Sorry for my slow response to this and other issues.  Things have been a bit crazy at the University of Washington and maintaining Able Player has taken a back seat to other priorities.  I'm planning to shift my focus back to Able Player starting next week (Monday May 22).  Stay tuned for updates...

**Comment by drumwhacker** [comment]
Thank you for the update, Terrill!  I'll suspend my investigation then (I'm
a software dev at Rio Salado College) and wait for your response first.

-Claire

**Comment by terrill** [comment]
Can someone provide a URL to a page that demonstrates this problem?  I'm unable to replicate it. From my tests, the Able Player chrome is built without error for YouTube videos that have no captions as well as those that have machine-generated captions only.  Also, be sure you're using the latest version of Able Player from the [develop branch](https://github.com/ableplayer/ableplayer/tree/develop).

**Comment by drumwhacker** [comment]
Hi Terrill, You can view what happens for me here:
https://riosalado.coursearc.com/content/toolbox/cg-test-blocks/ableplayer-v1-set-up-no-caption-track/

We have a home-rolled script that sets up ableplayer within a page multiple times based on class cues found.  Currently using 4.4.67, however, my tests use 4.5, which set up with the same problem (no control bar, sometimes autoplays).

**Comment by stuprice** [comment]
Hi

I'm suffering from this problem too and the link @drumwhacker provided above shows the exact same behaviour as I see.  I've compared the latest dev build to the release version and there are no changes in the area where I think the problem lies.

I have tried to debug the problem and here's what I think the issue is.

In `AblePlayer.prototype.getYouTubeCaptionTracks()` it determines there are no options for 'captions' or 'tracklist' so it's adds an event handler like this:

`// no tracks were found, probably because the captions module hasn't loaded
// play video briefly (required in order to load the captions module)
// and after the apiChange event is triggered, try again to retreive tracks
this.youTubePlayer.addEventListener('onApiChange',function(x)`

However, as there are no captions in the video, the captions module never loads so `onApiChange` never gets raised.  This means the video plays, doesn't get paused and `deferred.resolve();` never gets called.  

The impact is the chain of `If` statements in `AblePlayer.prototype.recreatePlayer = function ()` stops at `thisObj.getTracks().then(function()` and doesn't cascade down to create the player controls (as well as other stuff).

It seems we need to know that the event handler has been added and the video has started playing but the `apiOnChange` event hasn't fired and is never going to.  Alternatively if the captions state is unknown for a given period of time just give up and resolve the promise.

I hope this helps and makes sense.

**Comment by drumwhacker** [comment]
Hello Stu,
This makes sense. This would explain why the video keeps playing even
though it is not supposed to (auto) play by default.
There is a script I've placed in similar situations where, after a certain
number of tries, the func resolves or fails. I think I'll give that a shot.
-drum

**Comment by drumwhacker** [comment]
If you hard-refresh the test page I shared, you can see in the code where I have set up in the area you pointed out.  The timeout is arbitrary, probably could be a lot short like 500 or 300.

**Comment by stuprice** [comment]
Hi

I can see your page no longer autoplays the video but it still doesn't show the player controls so we're 50% of the way there.

However I can't see any of the changes you've made to make this happen?  Can you give me a pointer?

I'll try and have a go today to fix the ableplayer source when captions are missing.

Cheers

Stu

**Comment by drumwhacker** [comment]
Hi Stu,
In my view, the controls build out.
Here is the code I placed right under the line where ableplayer plays the video to try and trigger loading the captions module:

`let delayedPlayer = this.youTubePlayer;
setTimeout(stopVideo, 300);

function stopVideo() {

delayedPlayer.stopVideo();

let options = delayedPlayer.getOption('captions','tracklist');

//check if options is falsey
if(typeof options==='undefined'){
	console.log('RESOLVE SO IT SETS UP CONTROLS WITHOUT CAPTIONS');
	deferred.resolve();
}

`
It works on my side but I'll check again.

**Comment by stuprice** [comment]
Hi Claire

If I go into the dev tools in Chrome and Ctrl + Shift + F to search everywhere for a line of your code I can't find it:
`let options = delayedPlayer.getOption('captions','tracklist');`

I'm using the same page as you originally posted
[https://riosalado.coursearc.com/content/toolbox/cg-test-blocks/ableplayer-v1-set-up-no-caption-track/](https://riosalado.coursearc.com/content/toolbox/cg-test-blocks/ableplayer-v1-set-up-no-caption-track/)

I have just tested this code which works for me.  It's basically the same as yours except instead of waiting for an arbitrary amount of time it uses the `playing` event so we're sure the captions are not present.

			var onStateChangeEvent = this.youTubePlayer.addEventListener('onStateChange', function (x) {
				console.log(x);
				if (x.data === 1) { // Playing.
					thisObj.youTubePlayer.removeEventListener('onStateChange', onStateChangeEvent);

					if (!thisObj.youTubePlayer.getOption('captions', 'tracklist')) {
						if (!thisObj.okToPlay) {
							// Don't stopVideo() - that cancels loading 
							// Just pause 
							// No need to seekTo(0) - so little time has passed it isn't noticeable to the user 
							thisObj.youTubePlayer.pauseVideo();
						}

						// If we make it to here, the video has started playing but we have no captions
						// so give up on this function as we're never going to get them.
						deferred.resolve();
					}
				}
			});

Hope this helps :-)

**Comment by drumwhacker** [comment]
Hmmm,
If I do the same with Dev Tools, pull up the Sources tab and then pull up the Find bar, I DO find that line of code.  You definitely would have to do a hard refresh.  We keep our code on github but the upload of changes using Github Desktop is sometimes frustratingly slow for me.

I like your approach.  I will test it out here.

**Comment by stuprice** [comment]
Ok great!

I've just realised I should've used the constant YT.PlayerState.PLAYING instead of a literal 1 on this line:
`if (x.data === 1) { // Playing.`

Let me know how you get on.  I've done some more testing here and it seems fine for videos with & without captions.  I also sometimes have to use a VPN which is slow and emphasises the delays and load times which sort of helps with testing. 

Let me know how you get on, it would be nice to see this make its way into a build if we can.

**Comment by drumwhacker** [comment]
Hello Stu,
This code works for me if i include x.data===3 in the conditional.  Apparantly the player never hits PLAYING in my situation.  (state change Event Listener is added before player is asked to play to trigger the captions module load).

**Comment by drumwhacker** [comment]
I'm going to keep testing, on to multiple players.  My concern being that if I only get and respond to BUFFERING state, that's not enough to even allow a possible captions module to load, if there is one.  My situation might need a bit more time to get to an actual PLAYING state, so I may need to setTimeout...

**Comment by stuprice** [comment]
I did find I missed some events if I had breakpoints in the code so I had to disable them and rely on console logging to see what was going on.  

I've tried it with you video and it works ok:
[https://youtu.be/K04Rh1Jvfyc](https://youtu.be/K04Rh1Jvfyc)

This is the console output:
`ableplayer.js:6286 {target: cq, data: -1}`
`ableplayer.js:6286 {target: cq, data: 3}`
`ableplayer.js:6286 {target: cq, data: 1}`
`ableplayer.js:6286 {target: cq, data: 2}`

Do you still have your timeout running which is stopping the video before it has chance to play?

**Comment by drumwhacker** [comment]
I do not have my timeout running; I have commented out my code, using just yours.  Only change was to add the x.data===3 in the conditional because my player is not reaching the Playing state.  On my multiple players page, it's the same, I get:
data: -1
data: 3
(player is then paused, which is not a good look as the wheel keeps spinning in this state)
Have a few more tests.

**Comment by drumwhacker** [comment]
Update: After running a particular test video involving closed-captioning provided by the youtube video creator, it seems both of our solutions don't work on my end - our code prevents the loading of the captions (and creation of CC button) from the youtube side, which ableplayer does normally do. 
Auto-generated captions from you-tube do not get pulled in BUT they do trigger the captions module load, thus triggering the onAPIChange func.
Going to look into how to create/find the point where/when the code can get these captions but still have a point where the deferred var is resolved for absolutely *no* captions.

If you want to check that out on your end, this is the youtube video I used for this test which contains multiple language captions and one english auto-generated track which does not get pulled in:
[5MgBikgcWnY](https://www.youtube.com/watch?v=5MgBikgcWnY)

**Comment by drumwhacker** [comment]
Hello all,
The tldr:
I've looked into this further and after some experimentation and bringing other folks' thoughts into the issue, I've concluded that the use of timeOut seems necessary in order to both allow those yt videos that have yt captions to trigger the onApiChange event (and eventually resolve the promise) AND allow those yt videos that will *not* trigger the load of the yt caption module to resolve the promise and continue to get an ableplayer controls section.

Further reading:
**Code in Use**
```
var thisObj, ytTracks, i, trackLang, trackLabel, isDefaultTrack;
//ADDED: this flag for onApiChange event, changed to true inside the evt
var apiTriggered = false;

//ADDED: setTimeout placed below playVideo()
// Trigger the above event listener by briefly playing the video	
	this.loadingYouTubeCaptions = true;	
	this.youTubePlayer.playVideo();
			

  setTimeout(() => {
	if(promise.state() === 'pending' && !apiTriggered){
		
			setTimeout(() => {
				// Code to be executed after a delay
				console.log('ableplayer 4.5: Resolving after second timeout.');
				thisObj.youTubePlayer.stopVideo();
				deferred.resolve();
			}, 500);
	}
  },500);

```


**Test pages:**
1. Multiple players -captions not provided by youtube but our own interface. Players 1 and 3 should receive the captions button, Players 2 and 4 should not.
https://riosalado.coursearc.com/content/toolbox/cg-test-blocks/testing-ableplayer-interface-v4-5

2. Single player, receives captions (25 of them(!), from youtube user)
https://riosalado.coursearc.com/content/toolbox/cg-test-blocks/ableplayer-v1-set-up-no-caption-track

**Notes:**
I tried using setImmediate, a method used mostly in Node.js but has limited availability in browsers, just out of curiosity but
unfortunately it responds too quickly to allow the video with YT captions to trigger the onApiChange event.


Would appreciate any feedback on these testpages or using this setup in your own test scenarios!

**Comment by stuprice** [comment]
Hi Claire

Thanks for the update.  I have tried this update in my scenario with just a single video on screen with no captions and it works perfectly!  I have slight reservations about delaying for an arbitrary amount of time but it seems like we're out of all other options so this is great.

I have tested your two links above and my results do not quite match for Player 4.  It's probably not that important but I am running Window 11 with the latest versions of both Chrome and Edge and the results were the same in each browser.

#### 1. Multiple players
- Player 1 controls are visible and captions button active and clickable :-)
- Player 2 controls are visible and captions button inactive and not clickable :-)
- Player 3 controls are visible and captions button active and clickable which shows menu of languages :-)
- Player 4 **no controls are visible** :-(

#### 2. Single Player
- Player controls are visible and captions button active and clickable which shows menu of all 25 languages! :-)

This is the log from the Chrome Dev Tools which I captured on a page refresh.
[riosalado.coursearc.com-1689163586099.log](https://github.com/ableplayer/ableplayer/files/12027680/riosalado.coursearc.com-1689163586099.log)

**Comment by drumwhacker** [comment]
@stuprice Thank you very much for your testing.  I'm so sorry but I did not have my version of the player being tested set up for more than my team to view!  I will set up a different test setup and update you here.  My apologies again.

**Comment by stuprice** [comment]
Ah no worries! I did wonder why I couldn't find the able player changes in the dev tools but put it down to me missing something obvious :-)

Anyway, I made the changes locally and it worked for me.

**Comment by drumwhacker** [comment]
@stuprice, I'm glad the changes work for your scenario!  

I am looking into setting up a branch so that my test pages can be reviewed with all the correct files in place. It definitely wasn't you, lol, I'm glad you included your log.  I'll post here when I think I have a go.

**Comment by drumwhacker** [comment]
Hello all again (with updated URLS)

The tldr:
I've looked into this further and after some experimentation and bringing other folks' thoughts into the issue, I've concluded that the use of timeOut seems necessary in order to both allow those yt videos that have yt captions to trigger the onApiChange event (and eventually resolve the promise) AND allow those yt videos that will *not* trigger the load of the yt caption module to resolve the promise and continue to get an ableplayer controls section.

Further reading:
**Code in Use**
```
var thisObj, ytTracks, i, trackLang, trackLabel, isDefaultTrack;
//ADDED: this flag for onApiChange event, changed to true inside the evt
var apiTriggered = false;

//ADDED: setTimeout placed below playVideo()
// Trigger the above event listener by briefly playing the video	
	this.loadingYouTubeCaptions = true;	
	this.youTubePlayer.playVideo();
			

  setTimeout(() => {
	if(promise.state() === 'pending' && !apiTriggered){
		
			setTimeout(() => {
				// Code to be executed after a delay
				console.log('ableplayer 4.5: Resolving after second timeout.');
				thisObj.youTubePlayer.stopVideo();
				deferred.resolve();
			}, 500);
	}
  },500);

```


**Test pages:**
1. Multiple players -some captions not provided by youtube but our own interface. Players 1 and 3 should receive the captions button, Players 2 and 4 should not. **_Player 4 is the problem scenario.  Shown on a page with multiple players as the timing of the page load was a question to me._**
**NEW**
**https://riosalado.coursearc.com/content/toolbox/cg-test-blocks/ableplayer-interface-4-5-with-ableplayer-4-5-githubtest**

2. Single player, receives captions (25 of them(!), from youtube user)
**NEW**
**https://riosalado.coursearc.com/content/toolbox/cg-test-blocks/ableplayer-interface-4-5-with-ableplayer-4-5-githubtest-youtube-caps**

**Notes:**
I tried using setImmediate, a method used mostly in Node.js but has limited availability in browsers, just out of curiosity but
unfortunately it responds too quickly to allow the video with YT captions to trigger the onApiChange event.

Kept trying to use the other api offered from youtube, onStateChange, as suggested by @stuprice; however, for myself and
another coworker, we could not get to a point where a page with multiple players would consistently register 1 (Playing), 
instead getting -1(stopped) and 3(buffering), with the last player of the four not resolving and ergo, not getting a control bar.


**_Please check out these new page URLs as they contain the solution above for public viewing. My apologies for the previous URLS as they containded test components not available to the public._**

Would appreciate any feedback on these testpages or using this setup in your own test scenarios!

**Comment by cscottjames** [comment]
Bumping this - it still seems to be an issue.

**Comment by drumwhacker** [comment]
Hello,
I'm working on other projecrs right now but i do plan to come back and pick
this back up again.

**Comment by joedolson** [comment]
@drumwhacker I added a PR with what I understand you to be doing linked to this issue. Let me know if I've understood correctly.

I did change the call to `.stopVideo()` to `.pauseVideo()`, as stopVideo cancels loading.

What I'm finding is that it works part of the time; it's just not 100% dependable for me.

**Comment by joedolson** [comment]
OK. Fixed the PR; it's based against the develop branch.

**Comment by joedolson** [comment]
I found another way to possibly solve this. There's an undocumented path to use `setOption` to set a caption file, and if there are no captions available in that language, it will enable the auto captions generated by YouTube. 

This does two things: 1) it would make craptions available if no captions are provided and 2) *might* mean that this error with uncaptioned videos doesn't happen.

I did some tests, and it does work in a basic scenario, but there's a fair amount more to do to make it work properly.

However, to test it at a basic level, you can add 

```
if (typeof thisObj.youTubePlayer.setOption === 'function') {
	thisObj.youTubePlayer.setOption('captions', 'track', {languageCode: 'en'})
}
```

in the `else` block if there are no `ytTrack` in `getYouTubeCaptionTracks()`. 

Also need to set `thisObj.usingYouTubeCaptions = false;` and `thisObj.hasCaptions = false;` to `true` in that condition. 

I'm not going to explore this further until I start prepping 4.8, but this might be an alternate path.

**Comment by cscottjames** [comment]
Sadly, I'm still seeing this on develop and 4.7. I'm testing using a YouTube video with no words whatsoever. Examples are easy to find, like this for example https://www.youtube.com/watch?v=1x28RF72xZc
